lua/nvim-custom-diagnostic-highlight.lua

```diff
@@ -5,6 +5,7 @@
 ---@tag nvim-custom-diagnostic-highlight
 
 local nvim_custom_diagnostic_highlight = {}
+local augroup = vim.api.nvim_create_augroup('NvimCustomDiagnosticHighlight', { clear = true })
```

Comment:
I'd prefer to declare the augroup on `setup`, because people might want to run `setup` multiple times for different cases, with different highlight groups and whatnot.

Reply:
I can move it to `setup` but I think there's not much sense, because this would still be the same augroup all the time (because the name is the same). And this PR doesn't really care about the augroup, all autocommands are self-deleting or are deleted in the `hide` handler. 

If we put `nvim_create_augroup` in `setup` it won't change anything in most cases, e.g. when user calls `setup` 3 times in their config this means that we recreate and clear this augroup 3 times, then during runtime all autocommands are still using the same group. In fact in some (*very* unlikely) cases where user would *somehow* call `setup` during runtime when there is already some highlight deferred, it could cause errors, because `nvim_create_augroup(... { clear=true })` will delete all our autocommands, so the `hide` handler will try to delete non-existing commands.

Reply:
I was thinking to move that in `setup` and add something to the name, like `handler_name`, so that there would be a different augroup for each `setup` call.

Correct me if I'm wrong, but if we clear the augroup it would be recreated with only the configuration for that `setup` call. I'm thinking about the use case where someone has, for example, a `setup` call to highlight unused variables and another to highlight deprecated variables. In this case I think the best way to handle it would be to have a separate augroup (with different names) for each.

Reply:
Adding `handler_name` to augroup name should work fine as it is (as I understand it) unique to each `setup` call. So we can put e.g. `vim.api.nvim_create_augroup('NvimCustomDiagnosticHighlight_' .. handler_name, { clear = true })` in `setup`.

> Correct me if I'm wrong, but if we clear (...)

As it is currently (current version of this PR) everything should work fine, because augroup is created in top-level of the file, and `require('nvim-custom-diagnostic-highlight')` in Lua will execute the file only once. Problems might arise when we move `nvim_create_augroup(..., { clear = true })` to some function, e.g. `setup` because user might call it more than once. But if we add `handler_name` to augroup name it should work correctly as long as `handler_name` is unique for each `setup`. 

In theory we could use Nvim augroups to avoid all the logic of storing autocmds in `user_data` by actually putting `nvim_create_augroup` inside `show` but store handler_name, namespace and bufnr in augroup name, something like:
```lua
local augroup = vim.api.nvim_create_augroup(
  string.format('NvimCustomDiagnosticHighlight_%s_%s_%d', handler_name, namespace, bufnr),
  { clear = true })
```
Then we would just clear the augroup in `hide`. But I am not sure if it's not abusing augroups (but maybe it's better, I have no idea).

Reply:
I think using the `diagnostic_handler_namespace` would actually be better than the handler name, because the handler name is supposed to be ignored when `register_handler` is false. And `diagnostic_handler_namespace` must always exist and be unique, so I think it makes sense.

I understand what you mean by using the require only once, because the lua interpreter will cache the require result and won't re-run it, but I would prefer not to rely on this detail for the plugin not to break.

Reply:
Then `diagnostic_handler_namespace` seems like a better choice.

I'm not really sure how clearing the augroup could break the plugin.
Regarding the caching of `require` - I don't think it's relying on a detail, but rather on fundamental concept of how Lua works. It's not possible in Lua to re-run module's code, unless you explicitly unload the module from `package.loaded`. But doing so means that someone really wants to restart this module, so I think that clearing the augroup in such case is something user wants.

But anyway, I feel like we're overthinking this. In fact we could just use `clear = false`, because the created autocommands delete themselves or are deleted by `hide`, so clearing the namespace is not really necessary.

Reply:
Indeed. I did a bunch of tests with different calls to setup and all and it all seems to work well.

I believe I was overthinking the way the augroups are cleared, sorry about that.

I'll resolve this and merge. Thanks a lot for the contribuition! :rocket: